### PR TITLE
[SPARK-18686][SparkR][ML] Several cleanup and improvements for spark.logit.

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?
Several cleanup and improvements for ```spark.logit```:
* ```summary``` should return coefficients matrix, and should output labels for each class if the model is multinomial logistic regression model.
* ```summary``` should not return ```areaUnderROC, roc, pr, ...```, since most of them are DataFrame which are less important for R users. Meanwhile, these metrics ignore instance weights (setting all to 1.0) which will be changed in later Spark version. In case it will introduce breaking changes, we do not expose them currently.
* SparkR test improvement: comparing the training result with native R glmnet.
* Remove argument ```aggregationDepth``` from ```spark.logit```, since it's an expert Param(related with Spark architecture and job execution) that would be used rarely by R users.

## How was this patch tested?
Unit tests.

The ```summary``` output after this change:
multinomial logistic regression:
```
> df <- suppressWarnings(createDataFrame(iris))
> model <- spark.logit(df, Species ~ ., regParam = 0.5)
> summary(model)
$coefficients
             versicolor  virginica   setosa
(Intercept)  1.514031    -2.609108   1.095077
Sepal_Length 0.02511006  0.2649821   -0.2900921
Sepal_Width  -0.5291215  -0.02016446 0.549286
Petal_Length 0.03647411  0.1544119   -0.190886
Petal_Width  0.000236092 0.4195804   -0.4198165
```
binomial logistic regression:
```
> df <- suppressWarnings(createDataFrame(iris))
> training <- df[df$Species %in% c("versicolor", "virginica"), ]
> model <- spark.logit(training, Species ~ ., regParam = 0.5)
> summary(model)
$coefficients
             Estimate
(Intercept)  -6.053815
Sepal_Length 0.2449379
Sepal_Width  0.1648321
Petal_Length 0.4730718
Petal_Width  1.031947
```

### CODE REVIEW

**Comment by yanboliang** [comment]
cc @wangmiao1981 @felixcheung @jkbradley

File: mllib/src/main/scala/org/apache/spark/ml/r/LogisticRegressionWrapper.scala [hunk 1]
```diff
@@ -164,9 +174,10 @@ private[r] object LogisticRegressionWrapper
...
       val rMetadata = parse(rMetadataStr)
       val features = (rMetadata \ "features").extract[Array[String]]
+      val labels = (rMetadata \ "labels").extract[Array[String]]
 
       val pipeline = PipelineModel.load(pipelinePath)
-      new LogisticRegressionWrapper(pipeline, features, isLoaded = true)
+      new LogisticRegressionWrapper(pipeline, features, labels)
     }
   }
 }
```
Comment (wangmiao1981) [inline mllib/src/main/scala/org/apache/spark/ml/r/LogisticRegressionWrapper.scala]: new line?

File: R/pkg/R/mllib.R [hunk 2]
```diff
@@ -733,8 +733,6 @@ setMethod("predict", signature(object = "KMeansModel"),
 #'                  excepting that at most one value may be 0. The class with largest value p/t is predicted, where p
 #'                  is the original probability of that class and t is the class's threshold.
 #' @param weightCol The weight column name.
-#' @param aggregationDepth depth for treeAggregate (>= 2). If the dimensions of features or the number of partitions
-#'                         are large, this param could be adjusted to a larger size.
```
Comment (wangmiao1981) [inline R/pkg/R/mllib.R]: `aggregationDepth` is still on scala side. Any reason to remove it?

File: R/pkg/R/mllib.R [hunk 3]
```diff
@@ -746,45 +744,35 @@ setMethod("predict", signature(object = "KMeansModel"),
...
+#' training <- df[df$Species %in% c("versicolor", "virginica"), ]
+#' model <- spark.logit(training, Species ~ ., regParam = 0.5)
+#' summary <- summary(model)
+#'
+#' # fitted values on training data
+#' fitted <- predict(model, training)
+#'
 #' # save fitted model to input path
 #' path <- "path/to/model"
 #' write.ml(blr_model, path)
```
Comment (wangmiao1981) [inline R/pkg/R/mllib.R]: Since you changed `blr_model` -> `model`, it should be `model` here.

File: R/pkg/R/mllib.R [hunk 4]
```diff
@@ -817,44 +804,29 @@ setMethod("predict", signature(object = "LogisticRegressionModel"),
...
           function(object) {
             jobj <- object@jobj
-            is.loaded <- callJMethod(jobj, "isLoaded")
-
-            if (is.loaded) {
-              stop("Loaded model doesn't have training summary.")
+            features <- callJMethod(jobj, "rFeatures")
+            labels <- callJMethod(jobj, "labels")
+            coefficients <- callJMethod(jobj, "rCoefficients")
+            nCol <- length(coefficients) / length(features)
```
Comment (wangmiao1981) [inline R/pkg/R/mllib.R]: Can we do the `nCol` calculation and column name on scala side? So, we don't have to call `rFeatures` and `labels` on R side, which makes the logic simpler.

**Review by wangmiao1981 — COMMENTED** [review]
In general, it looks good to me. I have some minor comments.

File: R/pkg/R/mllib.R (hunk 3 above)
Comment (yanboliang) [inline R/pkg/R/mllib.R]: Good catch, updated. Thanks.

File: R/pkg/R/mllib.R (hunk 4 above)
Comment (yanboliang) [inline R/pkg/R/mllib.R]: Yeah, we could. The reason I did this way is we may add more model statistics which are different for binomial and multinomial logistic regression later, so we need to distinguish them at R side in any way.

**Comment by felixcheung** [comment]
LGTM except a minor comment on example. thanks!

**Comment by yanboliang** [comment]
Since this will involve breaking change if we merge it after 2.1 release, and actually I think it belongs to the scope of QA, so I'll merge it into master and branch-2.1, thanks for all your reviewing.

**Pull request closed without merging**